ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -679,27 +680,15 @@ Operator genOPTree(ASTNode ast, PlannerContext plannerCtx) throws SemanticExcept
 
           // Determine if we should re-throw the exception OR if we try to mark plan as reAnayzeAST to retry
           // planning as non-CBO.
-          if (e instanceof CalciteSubquerySemanticException || e instanceof CalciteViewSemanticException
-              || e instanceof CalciteSubqueryRuntimeException) {
-            // Non-CBO path for CalciteSubquerySemanticException fails with completely different error
-            // and masks the original failure.
-            // Non-CBO path for CalciteViewSemanticException would fail in a similar way as CBO path.
-            throw new SemanticException(e);
-          }
-
-          boolean isHiveTest = conf.getBoolVar(ConfVars.HIVE_IN_TEST);
-          // At this point we retry with CBO off:
-          // 1) If this is not test mode (common case)
-          // 2) If we are in test mode and we are missing stats
-          // 3) if we are in test mode and a CalciteSemanticException is generated
-          reAnalyzeAST = (!isHiveTest || isMissingStats ||  e instanceof CalciteSemanticException);
```

Comment:
there is a "small" change here - the new `CONSERVATIVE` doesn't contain this `isMissingStats` ; `e instanceof CSE` part.

Reply:
quesiton is what we want to do when a plain `RuntimeException` is thrown? do we want to retry it with the non-cbo path or not?

Reply:
Many thanks for keeping an eye @kgyrtkirk !

In production, (now `CONSERVATIVE` mode) `isHiveTest` is `false` so `!isHiveTest` becomes `true` thus no matter what happens with `isMissingStats` and `CalciteSemanticException`, we fallback to legacy optimizer.

In tests, before these changes we fallaback to legacy when the problem is due to missing stats or CSE. I see this mostly as a convenient way to not update a big amount of existing tests which otherwise will fail. From my perspective the fact that CBO fails is not something that should be considered normal. When writting a test if we know that we cannot handle it with CBO then I think it would be better to explicitly disable CBO instead of relying on this fallback mechanism. After these changes, future tests might (depends on the kind of exception that they come with) fail if they are missing stats.

Now regarding the `RuntimeException`:

- in `ALWAYS` mode we fallback to legacy;
- in `CONSERVATIVE` mode we fallback to legacy (same as before);
- in `NEVER` mode we stop and fail the query;
- in `TEST` mode we stop and fail the query (slightly different than before where we could retry if `RuntimeException` was paired with a missing stats problem).

Reply:
thank you for the explanation - yes it will be good

about the `RuntimeException` stuff - I think in case we encounter something unknown; we should just stop the show.
But because the approach is to blacklist some exception types - and retry everything else. Which probably aligns better to what our users want most of the time - get the resultset one way or the other.

Reply:
Personally, I would prefer to make the `CONSERVATIVE` mode fail for `RuntimeException`. This would given an incentive to the client to report the unknown problem (which could be serious) and at the same time it is not blocking since they could switch temporarily to the more permissive `ALWAYS` mode. If others feel the same then we can go for the change otherwise we leave it as is.

Reply:
okay we could add a new mode to experiment with this - but I'm not sure how usefull that would be; we can start with whitelisting the case when column stats are missing - but I think that will not get us much further...probably enabling that mode will expose in which other cases we should fall back...